modules/navier_stokes/src/linearfvkernels/LinearFVTKESourceSink.C

```diff
@@ -0,0 +1,269 @@
+
+  params.addParam<std::vector<BoundaryName>>(
+      "walls", {}, "Boundaries that correspond to solid walls.");
+  params.addParam<bool>(
+      "linearized_model",
+      true,
+      "Boolean to determine if the problem should be use in a linear or nonlinear solve.");
+  MooseEnum wall_treatment("eq_newton eq_incremental eq_linearized neq", "neq");
+
+  params.addParam<MooseEnum>("wall_treatment",
+                             wall_treatment,
+                             "The method used for computing the wall functions "
+                             "'eq_newton', 'eq_incremental', 'eq_linearized', 'neq'");
+  params.addParam<Real>("C_mu", 0.09, "Coupled turbulent kinetic energy closure.");
+  params.addParam<Real>("C_pl", 10.0, "Production Limiter Constant Multiplier.");
+
+  return params;
+}
+
+LinearFVTKESourceSink::LinearFVTKESourceSink(const InputParameters & params)
+  : LinearFVElementalKernel(params),
+    _dim(_subproblem.mesh().dimension()),
+    _u_var(getFunctor<Real>("u")),
+    _v_var(params.isParamValid("v") ? &(getFunctor<Real>("v")) : nullptr),
+    _w_var(params.isParamValid("w") ? &(getFunctor<Real>("w")) : nullptr),
+    _epsilon(getFunctor<Real>(NS::TKED)),
+    _rho(getFunctor<Real>(NS::density)),
+    _mu(getFunctor<Real>(NS::mu)),
+    _mu_t(getFunctor<Real>(NS::mu_t)),
+    _wall_boundary_names(getParam<std::vector<BoundaryName>>("walls")),
+    _linearized_model(getParam<bool>("linearized_model")),
+    _wall_treatment(getParam<MooseEnum>("wall_treatment").getEnum<NS::WallTreatmentEnum>()),
+    _C_mu(getParam<Real>("C_mu")),
+    _C_pl(getParam<Real>("C_pl"))
+{
+  if (_dim >= 2 && !_v_var)
+    paramError("v", "In two or more dimensions, the v velocity must be supplied!");
+
+  if (_dim >= 3 && !_w_var)
+    paramError("w", "In three or more dimensions, the w velocity must be supplied!");
+}
+
+void
+LinearFVTKESourceSink::initialSetup()
+{
+  LinearFVElementalKernel::initialSetup();
+  NS::getWallBoundedElements(
+      _wall_boundary_names, _fe_problem, _subproblem, blockIDs(), _wall_bounded);
+  NS::getWallDistance(_wall_boundary_names, _fe_problem, _subproblem, blockIDs(), _dist);
+  NS::getElementFaceArgs(_wall_boundary_names, _fe_problem, _subproblem, blockIDs(), _face_infos);
+}
+
+Real
+LinearFVTKESourceSink::computeMatrixContribution()
+{
+  /*
+  Matrix contribution:
+  - Computes near-wall TKE destruction
+  - Computes bulk TKE destruction
+  */
+
+  // Useful variables
+  const auto state = determineState();
+  const auto elem_arg = makeElemArg(_current_elem_info->elem());
+  const Real rho = _rho(elem_arg, state);
+
+  if (_wall_bounded.find(_current_elem_info->elem()) != _wall_bounded.end())
+  {
+
+    // Useful variables
+    const Real mu = _mu(elem_arg, state);
+    const Real TKE = _var.getElemValue(*_current_elem_info, state);
+
+    // Assembly variables
+    Real destruction = 0.0;
+    Real tot_weight = 0.0;
+    std::vector<Real> y_plus_vec, velocity_grad_norm_vec;
```

Comment:
you should reserve a size for these two vectors for performance purposes rather than allocating dynamically later on

Reply:
Thanks. These vectors have one or two components. Reserving should make no difference or even slow down the process. Please advise